Tidy up Kubernetes watcher synchronization

joestringer opened the pull request:
Cilium should no longer register or wait to sync k8s resources that are disabled such as LRP.

See commits for more details.

joestringer committed:
k8s: Make k8s resource init and wait consistent

Previously, the list of resources to be initialized would be prepared in
k.EnableK8sWatcher(), but the logic to handle the wait for cache sync
could select a different set of resources.

This PR tidies these up into smaller functions divided by the
functionality they provide (core CNI or feature additions), and
consistently enables these using the pkg/option configurations.

In future we can potentially even remove unused CRDs from the
cluster and Cilium will not complain that they are unavailable; this
commit doesn't try to go quite that far yet.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
k8s: Refactor Endpoint{,Slice} initialization

Refactor this logic into a dedicated function to simplify later commits.
No functional changes.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -305,57 +373,17 @@ func (k *K8sWatcher) InitK8sSubsystem(ctx context.Context) <-chan struct{} {
 	go func() {
 		select {
 		case <-cachesSynced:
-			log.Info("All pre-existing resources related to policy have been received; continuing")
```
These log messages mentioning this is "related to policy" have been wrong for some time now since we wait for just about every type of resource, so I just dropped those bits from the messages.

aanm commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -379,75 +407,57 @@ func (k *K8sWatcher) EnableK8sWatcher(ctx context.Context) error {
+		default:
+			log.WithFields(logrus.Fields{
+				logfields.Resource: r,
+			}).Warning("Not listening for Kubernetes resource updates for unhandled type")
```
This should be a Fatal IMO

aanm reviewed (commented):
Only one small comment

errordeveloper commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -367,9 +395,9 @@ type WatcherConfiguration interface {
-func (k *K8sWatcher) EnableK8sWatcher(ctx context.Context) error {
+// EnableK8sWatcher watches for policy, services and endpoint changes on the
+// Kubernetes api server defined in the receiver's daemon k8sClient.
+func (k *K8sWatcher) EnableK8sWatcher(ctx context.Context, resources []string) error {
```
Any thoughts on whether it might make sense to pass a map of handler functions instead of string slice here?

errordeveloper replied:
It crossed my mind while reading this code, it might make it easier to read as opposed to long switch/case block, just thought we could discuss that :)

joestringer replied:
I briefly thought about it, it does sound tidier. Only thing is that the function prototypes are different for the different init functions. I wasn't sure if it would make sense to force them all to conform to the same signature just for this cleanup.

errordeveloper reviewed (approved):
LGTM!

joestringer committed:
k8s: Consolidate resource registration

Shift the declaration of the CRD resources used by the agent & other
applications into pkg/k8s/synced, then declare a mapping from that
canonical list of resources to the corresponding groups inside
pkg/k8s/watchers. This makes both pieces of code synchronized. If for
instance CiliumEndpoint CRDs are disabled, they are globally disabled
everywhere rather than just in part of the agent code.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
k8s: Only register CRDs for enabled resources

Some resources may be dynamically disabled by configuration options in
the agent / operator. It doesn't make sense to register these resources
into Kubernetes if they are disabled in the agent, so don't register
them in that case.

Signed-off-by: Joe Stringer <joe@cilium.io>
